### PR TITLE
Run sidecars with dynamically-determined user IDs

### PR DESCRIPTION
**Pull request by aslakknutsen**
The sidecar injector sets the securityContext.runAsUser field regardless if it is specified in the sidecar template or not. This is to allow the sidecar template to not include the field, which is required to ensure that istioctl kube-inject outputs manifests that pass the SCC admission controller (if the runAsUser is specified, the SCC controller will reject the pod, as the uid won't be in the proper range). The sidecar injector then injects the correct runAsUser id even if it isn't specified in the template.

The Openshift profile is updated to
* enable the incrementUUID flag
* overlay the default deployments to allow Openshift to default securityContext

------
Co-authored-by: Marko Lukša <marko.luksa@gmail.com>
Co-authored-by: Brian Avery <bavery@redhat.com>
Co-authored-by: rcernich <rcernich@redhat.com>
Co-authored-by: Jacek Ewertowski <jewertow@redhat.com>

-----

These changes remove the need for the any anyuid setup as described in the Istio Openshift install Docs; 
https://istio.io/latest/docs/setup/platform-setup/openshift/  (separate pr coming)

### CODE REVIEW

File: Makefile.core.mk [hunk 1]
```diff
@@ -284,7 +284,8 @@ refresh-goldens:
 		./pkg/bootstrap/... \
 		./pkg/kube/inject/... \
 		./pilot/pkg/security/authz/builder/... \
-		./cni/pkg/plugin/...
+		./cni/pkg/plugin/... \
+		./istioctl/cmd/...
```
Comment (bleggett) [inline Makefile.core.mk:288]:
```suggestion
		./istioctl/cmd/...
```

Why is this added here? 

(I confess I don't understand what 'goldens' are - I presume based on comments they're binaries that rely on codegen'd sources in which case this addition makes sense/I don't understand why it wasn't already there)

File: pkg/kube/inject/webhook.go [hunk 2]
```diff
@@ -994,6 +1001,77 @@ func (wh *Webhook) inject(ar *kube.AdmissionReview, path string) *kube.Admission
...
+			// we're injecting a normal pod (with app container and optional sidecar container)
+			// we set proxyUID to the main app container's UID incremented by 1 (require unique uid for iptables routes)
+			// we set proxyGID to the main app container's UID
+			proxyUID, err = getProxyUIDFromAnnotation(pod)
+			if err != nil {
+				log.Infof("Could not get proxyUID from annotation: %v", err)
+			}
+			if proxyUID == nil {
+				if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.RunAsUser != nil {
+					proxyUID = pointer.Int64Ptr(*pod.Spec.SecurityContext.RunAsUser + 1)
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:1045]: what if the user is already 65535, will we create an invalid config?

File: pkg/kube/inject/webhook.go [hunk 3]
```diff
@@ -53,6 +54,8 @@ import (
 	"istio.io/pkg/log"
 )
 
+const ProxyUIDAnnotation = "sidecar.istio.io/proxyUID"
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:57]: What is this used for?

File: pkg/kube/inject/webhook.go [hunk 4]
```diff
@@ -1163,3 +1243,50 @@ func handleError(message string) {
 	log.Errorf(message)
 	totalFailedInjections.Increment()
 }
+
+// modifies the pod to ensure that the proxy runs with the given proxyUID instead of 1337
+func replaceProxyRunAsUserID(pod *corev1.Pod, proxyUID *int64) {
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:1248]: I am somewhat uncomfortable with this. This is both extremely coupled the injection template logic, and we have been trying to move logic into the templates instead of post-processing the pod where possible.

however, I think I need answers to some of my other comments to make any recommendations on alternatives

**Review by howardjohn — COMMENTED** [review]
In general I would be happy to reduce the friction on OpenShift here but I don't really "get it". Some more explanation (for people that don't know OpenShift) of what we are solving and why would be very helpful

**Comment by therealmitchconnors** [comment]
I heard several questions from users about this at Kubecon EU.  It would be great to improve the OpenShift experience if we can do so without impacting other users.  Also, we should update the OpenShift documentation on istio.io to reflect this change when it merges.

**Comment by jwendell** [comment]
I'm working on a second version of this PR. Will submit it for review this week.
